drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,921 @@
+			return -ETIMEDOUT;
+		}
+		k_msleep(1);
+	}
+
+	i2c_omap_reg_set(cfg, I2C_SYSC, sysc);
+
+	data->bb_valid = 0;
+	return 0;
+}
+
+/**
+ * @brief Set the speed of the OMAP I2C controller.
+ *
+ * This function sets the speed of the OMAP I2C controller based on the
+ * specified speed parameter. The speed can be set to either Fast mode or
+ * Standard mode.
+ *
+ * @param dev The pointer to the device structure.
+ * @param speed The desired speed for the I2C controller.
+ *
+ * @return 0 on success, negative error code on failure.
+ */
+static int i2c_omap_set_speed(const struct device *dev, uint32_t speed)
+{
+	struct i2c_omap_data *data = dev->data;
+
+	/* If configured for High Speed */
+	switch (speed) {
+	case I2C_BITRATE_FAST + 1 ... UINT32_MAX:
+		return -ENOTSUP;
+	case I2C_BITRATE_FAST:
+		/* Fast mode */
+		data->speed_config = (struct i2c_omap_speed_config){
+			.pscstate = 9,
+			.scllstate = 7,
+			.sclhstate = 5,
+		};
+		break;
+	case I2C_BITRATE_STANDARD:
+		/* Standard mode */
+		data->speed_config = (struct i2c_omap_speed_config){
+			.pscstate = 23,
+			.scllstate = 13,
+			.sclhstate = 15,
+		};
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the OMAP I2C controller.
+ *
+ * This function initializes the OMAP I2C controller by setting the speed and
+ * performing any necessary initialization steps.
+ *
+ * @param dev Pointer to the device structure for the I2C controller.
+ * @return 0 if successful, negative error code otherwise.
+ */
+static int i2c_omap_init(const struct device *dev)
+{
+	struct i2c_omap_data *data = dev->data;
+	const struct i2c_omap_cfg *cfg = dev->config;
+
+	k_sem_init(&data->cmd_complete, 0, 1);
+
+	/* Set the speed for I2C */
+	if (i2c_omap_set_speed(dev, cfg->speed)) {
+		LOG_ERR("Failed to set speed");
+		return -ENOTSUP;
+	}
+
+	i2c_omap_init_ll(dev);
+
+	return 0;
+}
+
+/**
+ * @brief Configure the OMAP I2C controller with the specified device configuration.
+ *
+ * This function configures the OMAP I2C controller with the specified device configuration.
+ *
+ * @param dev The pointer to the device structure.
+ * @param dev_config The device configuration to be applied.
+ *
+ * @return 0 on success, negative error code on failure.
+ */
+static int i2c_omap_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_omap_cfg *cfg = dev->config;
+	uint32_t speed_cfg;
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		speed_cfg = I2C_BITRATE_STANDARD;
+		goto out;
+	case I2C_SPEED_FAST:
+		speed_cfg = I2C_BITRATE_FAST;
+		goto out;
+	default:
+		return -ENOTSUP;
+	}
+
+	if ((dev_config & I2C_MODE_CONTROLLER) != I2C_MODE_CONTROLLER) {
+		return -ENOTSUP;
+	}
+
+	if ((dev_config & I2C_MSG_ADDR_10_BITS) != I2C_MSG_ADDR_10_BITS) {
+		return -ENOTSUP;
+	}
+
+out:
+	/* Take the I2C module out of reset */
+	i2c_omap_reg_set(cfg, I2C_CON, 0);
+	i2c_omap_set_speed(dev, speed_cfg);
+	i2c_omap_reset(dev);
```

Comment:
what is the need for resetting the module when initializing/taking the module out of reset? I don't see corresponding reset in reference Linux driver.

Reply:
Removed